### PR TITLE
Add RETRO model for pretraining

### PR DESCRIPTION
**Pull request by yidong72**
# What does this PR do ?

Added RETRO model, pretraining script, Mock Dataset. 
Added CI test pretraining with Mock Dataset.

### CODE REVIEW

File: examples/nlp/language_modeling/megatron_retro_pretraining.py
Comment (ericharper) [inline examples/nlp/language_modeling/megatron_retro_pretraining.py:64]: Typo. Should be `MegatronRetrievalModel`

File: nemo/collections/nlp/data/language_modeling/megatron/retro_dataset.py
Comment (ericharper) [inline nemo/collections/nlp/data/language_modeling/megatron/retro_dataset.py:30]: Clean up comments.

File: nemo/collections/nlp/data/language_modeling/megatron/retro_dataset.py
Comment (ericharper) [inline nemo/collections/nlp/data/language_modeling/megatron/retro_dataset.py:46]: seed is not configurable?

File: nemo/collections/nlp/data/language_modeling/megatron/retro_dataset.py
Comment (ericharper) [inline nemo/collections/nlp/data/language_modeling/megatron/retro_dataset.py:47]: This could be set in other places as well?

Comment (ericharper) [reply nemo/collections/nlp/data/language_modeling/megatron/retro_dataset.py:47]: https://github.com/PyTorchLightning/pytorch-lightning/blob/0fe2ae74e919c8ef220a029cab2203572ca536a5/pytorch_lightning/utilities/seed.py#L75

Comment (ericharper) [reply nemo/collections/nlp/data/language_modeling/megatron/retro_dataset.py:47]: Won't this call overwrite the seed?

Comment (ericharper) [reply nemo/collections/nlp/data/language_modeling/megatron/retro_dataset.py:47]: Also, gets called here: https://github.com/NVIDIA/NeMo/blob/ddd87197e94ca23ae54e641dc7784e64c00a43d6/nemo/collections/nlp/modules/common/megatron/megatron_init.py#L120

File: nemo/collections/nlp/models/language_modeling/megatron_retrival_model.py
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_retrival_model.py:58]: Please change the name of this file to `megatron_retrieval_model.py` and the name of this class to `MegatronRetrievalModel`

File: nemo/collections/nlp/models/language_modeling/megatron_retrival_model.py
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_retrival_model.py:60]: Retrieval

File: nemo/collections/nlp/models/language_modeling/megatron_retrival_model.py
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_retrival_model.py:73]: remove comment

File: nemo/collections/nlp/models/language_modeling/megatron_retrival_model.py
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_retrival_model.py:401]: Probably better to comment this out until we support PP

File: tests/collections/nlp/test_retrieval_module.py
Comment (ericharper) [inline tests/collections/nlp/test_retrieval_module.py:157]: typo: `retrieval`

**Review by ericharper — CHANGES_REQUESTED** [review]
See comments.

File: examples/nlp/language_modeling/conf/megatron_retro_config.yaml
Comment (yidong72) [reply examples/nlp/language_modeling/conf/megatron_retro_config.yaml:101]: it works in the RETRO. I just copy pasted the old comment

File: nemo/collections/nlp/models/language_modeling/megatron_retrival_model.py
Comment (yidong72) [reply nemo/collections/nlp/models/language_modeling/megatron_retrival_model.py:58]: fixed

File: nemo/collections/nlp/models/language_modeling/megatron_retrival_model.py
Comment (yidong72) [reply nemo/collections/nlp/models/language_modeling/megatron_retrival_model.py:60]: fixed

File: nemo/collections/nlp/data/language_modeling/megatron/retro_dataset.py
Comment (yidong72) [reply nemo/collections/nlp/data/language_modeling/megatron/retro_dataset.py:46]: This is mock dataset. the seed number is used to model different batch for different worker. No need to configure this.

File: nemo/collections/nlp/data/language_modeling/megatron/retro_dataset.py
Comment (yidong72) [reply nemo/collections/nlp/data/language_modeling/megatron/retro_dataset.py:47]: yeah. It is only useful for testing. I will replace it with real RETRO dataset later so seed number has no effects

File: nemo/collections/nlp/models/language_modeling/megatron_retrival_model.py
Comment (yidong72) [reply nemo/collections/nlp/models/language_modeling/megatron_retrival_model.py:73]: removed

File: nemo/collections/nlp/models/language_modeling/megatron_retrival_model.py
Comment (yidong72) [reply nemo/collections/nlp/models/language_modeling/megatron_retrival_model.py:318]: cleaned

File: nemo/collections/nlp/models/language_modeling/megatron_retrival_model.py
Comment (yidong72) [reply nemo/collections/nlp/models/language_modeling/megatron_retrival_model.py:401]: cleaned up

File: tests/collections/nlp/test_retrieval_module.py
Comment (yidong72) [reply tests/collections/nlp/test_retrieval_module.py:157]: fixed

File: examples/nlp/language_modeling/conf/megatron_retro_config.yaml
Comment (ericharper) [reply examples/nlp/language_modeling/conf/megatron_retro_config.yaml:101]: Okay let's clean up that comment then

**Review by ericharper — DISMISSED** [review]
LGTM. Thanks for the changes! Please wait for @MaximumEntropy to review before merging.

File: examples/nlp/language_modeling/megatron_retro_pretraining.py
Comment (MaximumEntropy) [inline examples/nlp/language_modeling/megatron_retro_pretraining.py:35]: Does it not support megatron_amp_O2 yet?

File: nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py:73]: Can we pass this via kwargs? Hard to understand what True, True, True, True is.

File: nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py:79]: Be careful here, if you do support megatron_amp_o2, you need to add the HaflPrecisionPlugin in your training script.

File: nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py:115]: This looks a little dangerous. What if pad_id = 0? If not 0 will evaluate to True. This will add another pad_id no? Can we make the check more explicit using hasattr(self.tokenizer, "pad_id") or self.tokenizer.pad > 0 or something like that

File: nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py:216]: Why are you casting to float32 here?

File: nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py:282]: Same as above. Do we need to cast to fp32?

File: nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py:338]: If not using pipeline parallel, then you want to check explicitly to make sure that global batch is always equal to micro batch * data_parallel_world_size. Do you check for this anywhere?

File: nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py:419]: Given that you don't support PP. why is async grad allreduce False?

File: nemo/collections/nlp/modules/common/megatron/retrieval_token_level_encoder_decoder.py
Comment (MaximumEntropy) [inline nemo/collections/nlp/modules/common/megatron/retrieval_token_level_encoder_decoder.py:89]: Can we pass a tokenizer object directly and then access whatever extra tokens we want from it like pad, eos. In the future, this may make it easier than passing specific token_ids.

**Review by MaximumEntropy — CHANGES_REQUESTED** [review]
Few comments/questions.

File: examples/nlp/language_modeling/megatron_retro_pretraining.py
Comment (yidong72) [reply examples/nlp/language_modeling/megatron_retro_pretraining.py:35]: removed support of `megatron_amp_o2`

File: nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py
Comment (yidong72) [reply nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py:73]: fixed

File: nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py
Comment (yidong72) [reply nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py:79]: removed support of `megatron_amp_o2`

File: nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py
Comment (yidong72) [reply nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py:216]: as we discussed, leave as is

File: nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py
Comment (yidong72) [reply nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py:282]: as we discussed, leave as is

File: nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py
Comment (yidong72) [reply nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py:338]: use the `MegatronPretrainingSampler` sampler

File: nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py
Comment (yidong72) [reply nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py:419]: removed `configure_optimizers` function

File: nemo/collections/nlp/modules/common/megatron/retrieval_token_level_encoder_decoder.py
Comment (yidong72) [reply nemo/collections/nlp/modules/common/megatron/retrieval_token_level_encoder_decoder.py:89]: if we passing-in the tokenizer, it makes unit test difficult as we need to instantiate a tokenizer. Passing an id is light weighted.

File: examples/nlp/language_modeling/megatron_retro_pretraining.py
Comment (MaximumEntropy) [inline examples/nlp/language_modeling/megatron_retro_pretraining.py:35]: What happens if precision is bf16? Don't we need the MixedPrecisionPlugin?

File: nemo/collections/nlp/modules/common/megatron/retrieval_token_level_encoder_decoder.py
Comment (MaximumEntropy) [reply nemo/collections/nlp/modules/common/megatron/retrieval_token_level_encoder_decoder.py:89]: Instantiating a tokenizer is just one function call though. @ericharper wrote GPT unittests recently with tokenizer instantiation, maybe we could just re-use that part?

Comment (yidong72) [reply nemo/collections/nlp/modules/common/megatron/retrieval_token_level_encoder_decoder.py]: I used a mock tokenizer in the test. It works this way

File: examples/nlp/language_modeling/megatron_retro_pretraining.py
Comment (yidong72) [reply examples/nlp/language_modeling/megatron_retro_pretraining.py]: added the bf16.

File: nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py
Comment (yidong72) [reply nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py]: this is how I fix it.
```python
        # add pad special token
        if not hasattr(self.tokenizer, "pad_id"):
            self.tokenizer.add_special_tokens({'pad_token': '<pad>'})
        elif hasattr(self.tokenizer, "pad_id") and (self.tokenizer.pad_id is None or self.tokenizer.pad_id < 0):
            self.tokenizer.add_special_tokens({'pad_token': '<pad>'})

```

**Review by MaximumEntropy — DISMISSED** [review]
LGTM! Really nice PR.

File: Jenkinsfile
Comment (okuchaiev) [inline Jenkinsfile:2390]: do we really need 30 steps?
Can we use ``trainer.fast_dev_run=True`` instead?

File: examples/nlp/language_modeling/conf/megatron_retro_config.yaml
Comment (okuchaiev) [inline examples/nlp/language_modeling/conf/megatron_retro_config.yaml:1]: let's call this something other than megatron_retro. tiny_retro or test_retro, etc. since this is only for 2 gpus.

File: nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py
Comment (okuchaiev) [inline nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py:385]: can you add docstring? What kind of parameters you want to be returned here, trainable, all?

File: Jenkinsfile
Comment (yidong72) [reply Jenkinsfile:2390]: It is a tiny model and runs fast. I use steps 30 so it can continue from the previous checkpoint.

File: examples/nlp/language_modeling/conf/megatron_retro_config.yaml
Comment (yidong72) [reply examples/nlp/language_modeling/conf/megatron_retro_config.yaml:1]: changed the name

File: nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py
Comment (yidong72) [reply nemo/collections/nlp/models/language_modeling/megatron_retrieval_model.py:378]: added doc string and move it to the base class

**Review by okuchaiev — APPROVED** [review]

**Pull request merged**
